singularity-eos/base/root-finding-1d/root_finding.hpp

```diff
@@ -235,6 +237,71 @@ PORTABLE_INLINE_FUNCTION Status regula_falsi(const T &f, const Real ytarget,
   return status;
 }
 
+// solves for f(x,params) - ytarget = 0
+template <typename T>
+PORTABLE_INLINE_FUNCTION Status newton_raphson(const T &f, const Real ytarget,
+                                               const Real guess, const Real a,
+                                               const Real b, const Real ytol, Real &xroot,
+                                               const RootCounts *counts = nullptr,
+                                               const bool &verbose = false) {
+
+  constexpr int max_iter = NEWTON_RAPHSON_NITER_MAX;
+  Real _x = guess;
+  Real _xold = 0.0;
+  auto status = Status::SUCCESS;
+
+  Real yg;
+  Real dfunc;
+
+  int iter;
+
+  for (iter = 0; iter < max_iter; iter++) {
+    std::tie(yg, dfunc) = f(_x); // C++11 tuple unpacking
+
+    // check if we are converged already
+    if (std::abs(yg - ytarget) < (ytol * ytarget)) break;
+
+    // not converged; compute the next step
+    _xold = _x;
+    _x = _x - (yg - ytarget) / dfunc;
+
+    // check if we are out of bounds
+    if (_x <= a && _xold <= a) {
+      if (verbose) {
+        printf("newton_raphson out of bounds! %.14e %.14e %.14e %.14e\n", ytarget, guess,
+               a, b);
+      }
```

Comment:
This is fine for now, but I strongly suspect this can be made much more robust with, e.g., a line search, or even a simple bounds enforcement. NR is prone to overshooting, and simply forcing it back in the domain is often enough to get it to converge.

Reply:
I agree. There is most likely several ways to improve the convergence properties. For now the point is more to implement the NR as given by the reference implementation. I'm planning to conduct some more extensive benchmarking in the future to establish the performance vs. convergence in this context (since I suspect that this implementation is tuned for performance with just "good enough" convergence properties), but there is no timeline yet.